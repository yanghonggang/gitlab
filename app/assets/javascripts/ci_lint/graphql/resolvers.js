import axios from '~/lib/utils/axios_utils';

const resolvers = {
  Mutation: {
    lintCI: (_, { endpoint, content, dry_run }) => {
      return axios.post(endpoint, { content, dry_run }).then(({ data }) => ({
        valid: data.valid,
        errors: data.errors,
        warnings: data.warnings,
        jobs: data.jobs.map(job => {
          const only = job.only ? { refs: job.only.refs, __typename: 'CiLintJobOnlyPolicy' } : null;

          return {
            name: job.name,
            stage: job.stage,
            beforeScript: job.before_script,
            script: job.script,
            afterScript: job.after_script,
            tagList: job.tag_list,
            environment: job.environment,
            when: job.when,
            allowFailure: job.allow_failure,
            only,
            except: job.except,
            __typename: 'CiLintJob',
          };
        }),
        __typename: 'CiLintContent',
      }));
    },
  },
};

export default resolvers;
