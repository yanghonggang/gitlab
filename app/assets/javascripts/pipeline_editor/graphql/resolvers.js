import Api from '~/api';

export const resolvers = {
  Query: {
    blobContent(_, { projectPath, path, ref }) {
      return {
        __typename: 'BlobContent',
        rawData: Api.getRawFile(projectPath, path, { ref }).then(({ data }) => {
          return data;
        }),
      };
    },
  },
};

export default resolvers;
