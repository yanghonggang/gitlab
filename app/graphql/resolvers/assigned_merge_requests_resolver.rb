# frozen_string_literal: true

module Resolvers
  class AssignedMergeRequestsResolver < UserMergeRequestsResolverBase
    type ::Types::MergeRequestType.connection_type, null: true
    accept_author

    def user_role
      :assignee
    end
  end
end
