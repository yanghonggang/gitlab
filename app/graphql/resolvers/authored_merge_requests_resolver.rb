# frozen_string_literal: true

module Resolvers
  class AuthoredMergeRequestsResolver < UserMergeRequestsResolverBase
    type ::Types::MergeRequestType.connection_type, null: true
    accept_assignee

    def user_role
      :author
    end
  end
end
