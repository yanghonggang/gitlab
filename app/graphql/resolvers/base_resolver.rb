# frozen_string_literal: true

module Resolvers
  class BaseResolver < GraphQL::Schema::Resolver
    extend ::Gitlab::Utils::Override
    include ::Gitlab::Utils::StrongMemoize
    include ::Gitlab::Graphql::GlobalIDCompatibility

    argument_class ::Types::BaseArgument

    def self.single
      @single ||= Class.new(self) do
        def ready?(**args)
          ready, early_return = super
          [ready, select_result(early_return)]
        end

        def resolve(**args)
          select_result(super)
        end

        def single?
          true
        end

        def select_result(results)
          results&.first
        end
      end
    end

    def self.last
      @last ||= Class.new(self.single) do
        def select_result(results)
          results&.last
        end
      end
    end

    def self.complexity
      0
    end

    def self.resolver_complexity(args, child_complexity:)
      complexity = 1
      complexity += 1 if args[:sort]
      complexity += 5 if args[:search]

      complexity
    end

    def self.complexity_multiplier(args)
      # When fetching many items, additional complexity is added to the field
      # depending on how many items is fetched. For each item we add 1% of the
      # original complexity - this means that loading 100 items (our default
      # maxp_age_size limit) doubles the original complexity.
      #
      # Complexity is not increased when searching by specific ID(s), because
      # complexity difference is minimal in this case.
      [args[:iid], args[:iids]].any? ? 0 : 0.01
    end

    override :object
    def object
      super.tap do |obj|
        # If the field this resolver is used in is wrapped in a presenter, unwrap its subject
        break obj.subject if obj.is_a?(Gitlab::View::Presenter::Base)
      end
    end

    def synchronized_object
      strong_memoize(:synchronized_object) do
        case object
        when BatchLoader::GraphQL
          object.sync
        else
          object
        end
      end
    end

    def single?
      false
    end

    def current_user
      context[:current_user]
    end

    # Overridden in sub-classes (see .single, .last)
    def select_result(results)
      results
    end
  end
end
