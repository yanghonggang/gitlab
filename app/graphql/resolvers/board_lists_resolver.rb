# frozen_string_literal: true

module Resolvers
  class BoardListsResolver < BaseResolver
    include BoardIssueFilterable
    include Gitlab::Graphql::Authorize::AuthorizeResource

    type Types::BoardListType, null: true

    argument :id, Types::GlobalIDType[List],
             required: false,
             description: 'Find a list by its global ID'

    argument :issue_filters, Types::Boards::BoardIssueInputType,
             required: false,
             description: 'Filters applied when getting issue metadata in the board list'

    alias_method :board, :object

    def resolve(lookahead: nil, id: nil, issue_filters: {})
      authorize!(board)

      lists = board_lists(id)
      context.scoped_set!(:issue_filters, issue_filters(issue_filters))

      if load_preferences?(lookahead)
        List.preload_preferences_for_user(lists, context[:current_user])
      end

      Gitlab::Graphql::Pagination::OffsetActiveRecordRelationConnection.new(lists)
    end

    private

    def board_lists(id)
      service = ::Boards::Lists::ListService.new(
        board.resource_parent,
        context[:current_user],
        list_id: extract_list_id(id)
      )

      service.execute(board, create_default_lists: false)
    end

    def authorized_resource?(board)
      Ability.allowed?(context[:current_user], :read_list, board)
    end

    def load_preferences?(lookahead)
      lookahead&.selection(:edges)&.selection(:node)&.selects?(:collapsed)
    end

    def extract_list_id(gid)
      return unless gid.present?

      GitlabSchema.parse_gid(gid, expected_type: ::List).model_id
    end
  end
end
