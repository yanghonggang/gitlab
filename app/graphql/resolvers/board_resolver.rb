# frozen_string_literal: true

module Resolvers
  class BoardResolver < BaseResolver.single
    alias_method :parent, :synchronized_object

    type Types::BoardType, null: true

    argument :id, ::Types::GlobalIDType[::Board],
             required: true,
             description: 'The board\'s ID'

    def resolve(id: nil)
      return unless parent

      ::Boards::ListService.new(parent, context[:current_user], board_id: extract_board_id(id)).execute(create_default_board: false).first
    rescue ActiveRecord::RecordNotFound
      nil
    end

    private

    def extract_board_id(gid)
      GitlabSchema.parse_gid(gid, expected_type: ::Board).model_id
    end
  end
end
