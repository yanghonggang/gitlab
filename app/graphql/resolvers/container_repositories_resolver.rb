# frozen_string_literal: true

module Resolvers
  class ContainerRepositoriesResolver < BaseResolver
    type Types::ContainerRepositoryType, null: true

    argument :name, GraphQL::STRING_TYPE,
              required: false,
              description: 'Filter the container repositories by their name'

    def resolve(name: nil)
      ContainerRepositoriesFinder.new(user: current_user, subject: object, params: { name: name })
                                 .execute
                                 .tap { track_event(:list_repositories, :container) }
    end

    private

    def track_event(event, scope)
      ::Packages::CreateEventService.new(nil, current_user, event_name: event, scope: scope).execute
      ::Gitlab::Tracking.event(event.to_s, scope.to_s)
    end
  end
end
