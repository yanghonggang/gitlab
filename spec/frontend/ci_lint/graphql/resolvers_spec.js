import MockAdapter from 'axios-mock-adapter';
import axios from '~/lib/utils/axios_utils';
import httpStatus from '~/lib/utils/http_status';

import resolvers from '~/ci_lint/graphql/resolvers';
import { mockLintResponse } from '../mock_data';

describe('~/ci_lint/graphql/resolvers', () => {
  let mock;

  beforeEach(() => {
    mock = new MockAdapter(axios);
  });

  afterEach(() => {
    mock.restore();
  });

  describe('Mutation', () => {
    describe('lintCI', () => {
      const endpoint = '/ci/lint';

      beforeEach(() => {
        mock.onPost(endpoint).reply(httpStatus.OK, mockLintResponse);
      });

      it('resolves lint data with type names', async () => {
        const result = resolvers.Mutation.lintCI(null, {
          endpoint,
          content: 'content',
          dry_run: true,
        });

        await expect(result).resolves.toMatchSnapshot();
      });
    });
  });
});
