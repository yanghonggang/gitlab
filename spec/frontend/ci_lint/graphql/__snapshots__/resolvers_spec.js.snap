// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`~/ci_lint/graphql/resolvers Mutation lintCI resolves lint data with type names 1`] = `
Object {
  "__typename": "CiLintContent",
  "errors": Array [],
  "jobs": Array [
    Object {
      "__typename": "CiLintJob",
      "afterScript": Array [
        "echo 'after script 1",
      ],
      "allowFailure": false,
      "beforeScript": Array [
        "echo 'before script 1'",
      ],
      "environment": "prd",
      "except": Object {
        "refs": Array [
          "master@gitlab-org/gitlab",
          "/^release/.*$/@gitlab-org/gitlab",
        ],
      },
      "name": "job_1",
      "only": null,
      "script": Array [
        "echo 'script 1'",
      ],
      "stage": "test",
      "tagList": Array [
        "tag 1",
      ],
      "when": "on_success",
    },
    Object {
      "__typename": "CiLintJob",
      "afterScript": Array [
        "echo 'after script 2",
      ],
      "allowFailure": true,
      "beforeScript": Array [
        "echo 'before script 2'",
      ],
      "environment": "stg",
      "except": Object {
        "refs": Array [
          "master@gitlab-org/gitlab",
          "/^release/.*$/@gitlab-org/gitlab",
        ],
      },
      "name": "job_2",
      "only": Object {
        "__typename": "CiLintJobOnlyPolicy",
        "refs": Array [
          "web",
          "chat",
          "pushes",
        ],
      },
      "script": Array [
        "echo 'script 2'",
      ],
      "stage": "test",
      "tagList": Array [
        "tag 2",
      ],
      "when": "on_success",
    },
  ],
  "valid": true,
  "warnings": Array [],
}
`;
